libs/api/libbpf_program.cpp

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#include "api_internal.h"
+#pragma warning(push)
+#pragma warning(disable : 4200)
+#include "libbpf.h"
+#pragma warning(pop)
+#include "libbpf_internal.h"
+
+// This file implements APIs in LibBPF's bpf.h
+// and is based on code in libbpf.c.
+
+static const ebpf_program_type_t*
+_get_ebpf_program_type(enum bpf_prog_type type)
+{
+    // TODO(issue #223): read this mapping from the registry
+    switch (type) {
+    case BPF_PROG_TYPE_XDP:
+        return &EBPF_PROGRAM_TYPE_XDP;
+    }
+    return nullptr;
+}
+
+static const ebpf_attach_type_t*
+_get_ebpf_attach_type(enum bpf_attach_type type)
+{
+    // TODO(issue #223): read this mapping from the registry
+    switch (type) {
+    case BPF_ATTACH_TYPE_XDP:
+        return &EBPF_ATTACH_TYPE_XDP;
+    }
+    return nullptr;
+}
+
+static enum bpf_attach_type
+_get_bpf_attach_type(const ebpf_attach_type_t* type)
+{
+    // TODO(issue #223): read this mapping from the registry
+    if (memcmp(type, &EBPF_ATTACH_TYPE_XDP, sizeof(*type)) == 0) {
+        return BPF_ATTACH_TYPE_XDP;
+    }
+    return BPF_ATTACH_TYPE_UNKNOWN;
+}
+
+int
+bpf_prog_load(const char* file_name, enum bpf_prog_type type, struct bpf_object** object, int* program_fd)
+{
+    const ebpf_program_type_t* program_type = _get_ebpf_program_type(type);
+
+    if (program_type == nullptr) {
+        return EBPF_INVALID_ARGUMENT;
+    }
+
+    const char* log_buffer;
+    return (int)ebpf_program_load(
+        file_name, program_type, nullptr, EBPF_EXECUTION_ANY, object, (fd_t*)program_fd, &log_buffer);
+}
+
+int
+bpf_program__fd(const struct bpf_program* program)
+{
+    return (int)ebpf_program_get_fd(program);
+}
+
+const char*
+bpf_program__name(const struct bpf_program* program)
+{
+    return program->program_name;
+}
+
+const char*
+bpf_program__section_name(const struct bpf_program* program)
+{
+    return program->section_name;
+}
+
+size_t
+bpf_program__size(const struct bpf_program* program)
+{
+    return program->byte_code_size;
+}
+
+struct bpf_link*
+bpf_program__attach(struct bpf_program* program)
+{
+    if (program == nullptr) {
+        return nullptr;
+    }
+
+    ebpf_handle_t link_handle;
+    uint32_t result = ebpf_api_link_program(program->handle, program->attach_type, &link_handle);
+    if (result) {
+        errno = result;
+        return nullptr;
+    }
+
+    // The bpf_link structure is opaque so we can use the link handle directly.
+    return (struct bpf_link*)link_handle;
```

Comment:
This will need to return pointer to ebpf_link_t once the PR for issue #81 is completed.

Reply:
Agree.  Added comments referencing issue 81